engine/engine.go

```diff
@@ -196,125 +195,256 @@ func (r *ruleEngine) RunRulesScoped(ctx context.Context, ruleSets []RuleSet, sco
+// returns tags if there were any tagging rules in the batch
+// returns error if context was canceled during rule execution
+func (r *ruleEngine) runRulesBatch(
+	ctx context.Context,
+	batchName string,
+	totalRules int32,
+	rules []ruleMessage,
+	scopes Scope,
+	ruleContext ConditionContext,
+	carrier propagation.MapCarrier,
+	rulesetsMap map[string]*konveyor.RuleSet,
+) (map[string]interface{}, error) {
+	ruleResponseChan := make(chan response)
 
 	var matchedRules int32
 	var unmatchedRules int32
 	var failedRules int32
 
+	// tagging rules create new tag
+	allTags := map[string]interface{}{}
+	tagsChan := make(chan string)
+
 	wg := &sync.WaitGroup{}
+
+	// handle tags generated during handling of returns
+	go func() {
+		for {
+			select {
+			case tag := <-tagsChan:
+				allTags[tag] = true
+				wg.Done()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
 	// Handle returns
 	go func() {
 		for {
 			select {
-			case response := <-ret:
+			case response := <-ruleResponseChan:
 				func() {
 					r.logger.Info("rule returned", "ruleID", response.Rule.RuleID)
 					defer wg.Done()
 					if response.Err != nil {
 						atomic.AddInt32(&failedRules, 1)
 						r.logger.Error(response.Err, "failed to evaluate rule", "ruleID", response.Rule.RuleID)
 
-						if rs, ok := mapRuleSets[response.RuleSetName]; ok {
+						if rs, ok := rulesetsMap[response.RuleSetName]; ok {
 							rs.Errors[response.Rule.RuleID] = response.Err.Error()
 						}
 					} else if response.ConditionResponse.Matched && len(response.ConditionResponse.Incidents) > 0 {
+						r.logger.V(5).Info("rule matched", "ruleID", response.Rule.RuleID, "batch", batchName)
 						violation, err := r.createViolation(ctx, response.ConditionResponse, response.Rule, scopes)
 						if err != nil {
 							r.logger.Error(err, "unable to create violation from response", "ruleID", response.Rule.RuleID)
 						}
 						if len(violation.Incidents) == 0 {
-							r.logger.V(5).Info("rule was evaluated and incidents were filtered out to make it unmatched", "ruleID", response.Rule.RuleID)
+							r.logger.V(5).Info("rule was evaluated but all incidents were filtered out", "ruleID", response.Rule.RuleID)
 							atomic.AddInt32(&unmatchedRules, 1)
-							if rs, ok := mapRuleSets[response.RuleSetName]; ok {
+							if rs, ok := rulesetsMap[response.RuleSetName]; ok {
 								rs.Unmatched = append(rs.Unmatched, response.Rule.RuleID)
 							}
-						} else {
-							atomic.AddInt32(&matchedRules, 1)
-							rs, ok := mapRuleSets[response.RuleSetName]
-							if !ok {
-								r.logger.Info("this should never happen that we don't find the ruleset")
-								return
+							return
+						}
+						atomic.AddInt32(&matchedRules, 1)
+						rs, ok := rulesetsMap[response.RuleSetName]
+						if !ok {
+							r.logger.Info("this should never happen that we don't find the ruleset")
+							return
+						}
+						if response.Rule.Perform.Tag != nil {
+							// handle a tagging rule (info rule)
```

Comment:
Just a suggestion: I would move this to a different method, since it seems to me a more specific concern